File: BinarySearch/69_Sqrt_x.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-


class Solution(object):
    # Binary search.
    def mySqrt(self, x):
        low, high = 0, x
        while low <= high:
            mid = (low + high) / 2
            if mid ** 2 <= x < (mid + 1) ** 2:
                return mid
            elif mid ** 2 > x:
                high = mid - 1
            else:
                low = mid + 1


class Solution_2(object):
    # Newton iterative method
    # According to:
    # http://www.matrix67.com/blog/archives/361
    def mySqrt(self, x):
        if not x:
            return x
        val = x
        sqrt_x = (val + x * 1.0 / val) / 2.0
        while val - sqrt_x > 0.001:
            val = sqrt_x
            sqrt_x = (val + x * 1.0 / val) / 2.0

        return int(sqrt_x)


class Solution_3(object):
    # Shorter Newton method.
    def mySqrt(self, x):
        val = x
        while val * val > x:
            val = (val + x / val) / 2
        return val

"""
0
1
15
90
1010
"""

File: BinarySearch/69_Sqrt_x.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-07-01 20:42:32
 */

class Solution {
public:
    int mySqrt(int x) {
        if(x < 2){
            return x;
        }
        int low = 0, high = x;
        while(low <= high){
            int mid = low + (high - low) / 2;
            // mid * mid will overflow when mid > sqrt(INT_MAX)
            if(x / mid >= mid  && x / (mid+1) < (mid+1)){
                return mid;
            }
            else if(x / mid < mid){
                high = mid - 1;
            }
            else{
                low = mid + 1;
            }
        }
        return -1; // Can'e be reached!
    }
};

class Solution_2{
public:
    int mySqrt(int x) {
        // Newton iterative method
        long ans = x;
        while(ans * ans > x){
            ans = (ans + x/ans) / 2;
        }
        return ans;
    }
};

/*
0
1
15
90
1010
*/
